python/mujinasync/asynctcp.py

```diff
@@ -255,62 +275,103 @@ def _DestroyServerSocket(self):
 
-    def UnregisterServer(self, server):
+    def UnregisterServer(self, server: TcpServer):
         if server in self._servers:
             self._servers.remove(server)
 
-    def RegisterClient(self, client):
+    def RegisterClient(self, client: TcpClient):
         if client not in self._clients:
             self._clients.append(client)
 
-    def UnregisterClient(self, client):
+    def UnregisterClient(self, client: TcpClient):
         if client in self._clients:
             self._clients.remove(client)
 
+    def RegisterSocket(self, sock: socket.socket, mask: int) -> None:
+        """
+        Register socket to selector
+        Should be called when socket is created
+        """
+        assert self._selector, "selector is not ready"
+        existingMask = self._registeredSocketMaskBySocket.get(sock)
+        if existingMask == mask:
+            return
+
+        try:
+            if existingMask is None:
+                # register new socket to selector
+                self._selector.register(sock, mask, data=sock)
+                self._registeredSocketMaskBySocket[sock] = mask
+            else:
+                # update exisiting socket in selector
+                self._selector.modify(sock, mask, data=sock)
+                self._registeredSocketMaskBySocket[sock] = mask
+        except (OSError, ValueError) as e:
+            log.warning('failed to register socket %s: %s', sock, e)
+
+    def UnregisterSocket(self, sock: socket.socket) -> None:
+        """
+        Unregister socket from selector
+        Should be called when socket is destoryed
+        """
+        assert self._selector, "selector is not ready"
+        existingMask = self._registeredSocketMaskBySocket.get(sock)
+        if existingMask is None:
+            return
+
+        try:
+            self._selector.unregister(sock)
+            self._registeredSocketMaskBySocket.pop(sock)
+        except (OSError, ValueError, KeyError) as e:
+            log.warning('failed to unregister unused socket %s: %s', sock, e)
+
     def SpinOnce(self, timeout=0):
         """Spin all sockets once, without creating threads.
 
         :param timeout: in seconds, pass in 0 to not wait for socket events, otherwise, will wait up to specified timeout
         """
-        newConnections = [] # list of tuple (serverClient, connection)
-
-        # construct a list of connections to select on
-        rsockets = []
-        wsockets = []
-        xsockets = []
+        assert self._selector, "selector is not ready"
+        newConnections: list[tuple[TcpServerClient, TcpConnection]] = [] # list of tuple (serverClient, connection)
+        socketConnections: dict[socket.socket, tuple[TcpServerClient, TcpConnection]] = {} # Track socket->connection mapping and server sockets
 
         # bind and listen for server
-        serverSockets = {} # map from serverSocket to server
+        serverSockets: dict[socket.socket, TcpServer] = {} # map from serverSocket to server
```

Comment:
minor, but since this line is changing anyway I might convert `serverSockets` to something that describes what it is more accurately (e.g, `tcpServersBySocket`)